server/src/main/java/org/elasticsearch/index/mapper/DotExpandingXContentParser.java

```diff
@@ -58,37 +59,112 @@ public Token nextToken() throws IOException {
             if (token != Token.FIELD_NAME) {
                 return token;
             }
-            expandDots();
+            expandDots(delegate);
             return Token.FIELD_NAME;
         }
 
-        private void expandDots() throws IOException {
+        private void expandDots(XContentParser delegate) throws IOException {
             // this handles fields that belong to objects that can't hold subobjects, where the document specifies
             // the object holding the flat fields
             // e.g. { "metrics.service": { "time.max" : 10 } } with service having subobjects set to false
             if (contentPath.isWithinLeafObject()) {
                 return;
             }
-            XContentParser delegate = delegate();
             String field = delegate.currentName();
-            String[] subpaths = splitAndValidatePath(field);
-            // Corner case: if the input has a single trailing '.', eg 'field.', then we will get a single
-            // subpath due to the way String.split() works. We can only return fast here if this is not
-            // the case
-            // TODO make this case throw an error instead? https://github.com/elastic/elasticsearch/issues/28948
-            if (subpaths.length == 1 && field.endsWith(".") == false) {
+            int length = field.length();
+            if (length == 0) {
+                throw new IllegalArgumentException("field name cannot be an empty string");
+            }
+            final int dotCount = FieldTypeLookup.dotCount(field);
+            if (dotCount == 0) {
                 return;
             }
+            doExpandDots(delegate, field, dotCount);
+        }
+
+        private void doExpandDots(XContentParser delegate, String field, int dotCount) throws IOException {
```

Comment:
It seems like it would be faster to use a `StringTokenizer` with a `.` delimiter and switch to using an `ArrayList` for `subpaths`. 

Indeed, if copying to the `String[]` is the issue, why not remove the issue?

For other things `Arrays.asList` simply wraps the existing array, so there isn't any overhead there for things that output an `String[]` only.

Reply:
Not sure we should add new code using the tokenizer? While not deprecated its use is discouraged? 

Also, I think the use of any `ArrayList` here won't do us much good. That just adds more indirection and method calls. The main difference between this solution and `main` is probably in the better inlining, fewer allocations and more cache-locality.

main profiles like this:

![image](https://user-images.githubusercontent.com/6490959/223539356-cfd5d5f3-7b73-4336-9005-1e9b12720652.png)

while this branch profiles as:

![image](https://user-images.githubusercontent.com/6490959/223539478-be1b1f44-b15e-4c5a-95ae-3c277b03732b.png)

-> more inlining + no visible allocation pressure, there's no way the JVM will be able to escape analyse away an array list here for comparable performance I think